### PR TITLE
Removed Plugin Directory code from ExtensionsOrchestrator

### LINKED ISSUE #3696: [Feature/extensions] Remove plugin directory code from ExtensionsOrchestrator
Unnecessary code for reading plugin directories needs to be removed from ExtensionsOrchestrator.  Unit tests will also be restructured to reflect this change.

### PR DESCRIPTION
**Pull request by ryanbogan**
### Description
Removes plugin directory code from 
 
### Issues Resolved
#3696 
 
### Check List
- [X ] New functionality includes testing.
  - [X ] All tests pass
- [X ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 1]
```diff
@@ -86,18 +85,18 @@ public static enum RequestType {
...
     TransportService transportService;
     ClusterService clusterService;
 
     public ExtensionsOrchestrator(Settings settings, Path extensionsPath) throws IOException {
         logger.info("ExtensionsOrchestrator initialized");
         this.extensionsPath = extensionsPath;
         this.transportService = null;
-        this.extensionsSet = new HashSet<DiscoveryExtension>();
-        this.extensionsInitializedSet = new HashSet<DiscoveryExtension>();
-        this.clusterService = null;
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:100]: We need `clusterService` just like `transportService ` above

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:100]: Looks like I accidentally deleted an extra line

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 2]
```diff
@@ -150,56 +149,52 @@ private void extensionsDiscovery() throws IOException {
...
-                    throw new IOException("Could not load plugin descriptor " + plugin.getFileName(), e);
+                    extensionsList.add(
+                        new DiscoveryExtension(
+                            extension.getName(),
+                            extension.getUniqueId(),
+                            // placeholder for ephemeral id, will change with POC discovery
+                            extension.getUniqueId(),
+                            extension.getHostName(),
+                            extension.getHostAddress(),
+                            new TransportAddress(InetAddress.getByName(extension.getHostAddress()), Integer.parseInt(extension.getPort())),
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:170]: Should we keep it as `TransportAddress.META_ADDRESS`? @saratvemulapalli any thoughts here

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:170]: ^ we changed it this morning while troubleshooting the connection errors

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:170]: For additional context

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:170]: Yeah. I'm just trying to think of a better approach here

**Comment by owaiskazi19** [comment]
@peterzhuamazon do we have gradle check workflow enabled for feature branches?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 3]
```diff
@@ -150,56 +149,52 @@ private void extensionsDiscovery() throws IOException {
...
+                                Boolean.parseBoolean(extension.hasNativeController())
+                            )
+                        )
+                    );
+                    logger.info("Loaded extension: " + extension);
+                } catch (IllegalArgumentException e) {
+                    logger.error(e.toString());
                 }
             }
+            logger.info("Loaded all extensions");
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:190]: We should add a check here before logging the message. Case: If there's only one extension and even that gets caught in the catch block above.

**Comment by peterzhuamazon** [comment]
> @peterzhuamazon do we have gradle check workflow enabled for feature branches?

@owaiskazi19 seems like my initial assumption is not correct, at least the feature branch need a github actions yml to trigger.

Please let me know what are the branches I need to backport.

Thanks.

**Comment by owaiskazi19** [comment]
> Please let me know what are the branches I need to backport.

Can you check with core team? For extensibility we have `feature/extensions`

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 4]
```diff
@@ -86,18 +85,19 @@ public static enum RequestType {
     };
 
     private final Path extensionsPath;
-    final Set<DiscoveryExtension> extensionsSet;
-    Set<DiscoveryExtension> extensionsInitializedSet;
+    final List<DiscoveryExtension> extensionsList;
```
Comment (saratvemulapalli) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:88]: Curious why a List though?

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 5]
```diff
@@ -67,84 +66,79 @@ public void testExtensionsDiscovery() throws Exception {
             "     hostName: 'myIndependentPluginHost1'",
             "     hostAddress: '127.0.0.0'",
             "     port: '9300'",
-            "     version: '3.0.0'",
+            "     version: '1.0'",
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:69]: Are we really going to be so boring as to use 1.0 for a version for a test?  There are an infinite number of numbers out there.   0.0.7, for example. 3.14.16.  2.7.18.  42.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 6]
```diff
@@ -150,56 +150,56 @@ private void extensionsDiscovery() throws IOException {
...
 
-        Set<Extension> extensions = new HashSet<Extension>();
-        if (!pluginDirs.isEmpty()) {
+        List<Extension> extensions = new ArrayList<Extension>();
+        if (Files.exists(extensionsPath.resolve("extensions.yml"))) {
             try {
                 extensions = readFromExtensionsYml(extensionsPath.resolve("extensions.yml")).getExtensions();
             } catch (Exception e) {
-                throw new IOException("Could not read from extensions.yml");
+                throw new IOException("Could not read from extensions.yml\n " + e);
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:159]: Why not use [`IOException(message, throwable)`](https://docs.oracle.com/javase/7/docs/api/java/io/IOException.html#IOException(java.lang.String,%20java.lang.Throwable)) syntax here?  (Not sure if that's better, just asking.)

Also per earlier comment, if `readFromExtensionsYml` is throwing `IOException` what is the value add of catching it just to re-throw it?

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 7]
```diff
@@ -150,56 +150,56 @@ private void extensionsDiscovery() throws IOException {
...
         }
-        List<Path> pluginDirs = PluginsService.findPluginDirs(extensionsPath);
 
-        Set<Extension> extensions = new HashSet<Extension>();
-        if (!pluginDirs.isEmpty()) {
+        List<Extension> extensions = new ArrayList<Extension>();
+        if (Files.exists(extensionsPath.resolve("extensions.yml"))) {
             try {
                 extensions = readFromExtensionsYml(extensionsPath.resolve("extensions.yml")).getExtensions();
             } catch (Exception e) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:158]: Generally, catching all exceptions can hide other bugs.

`readFromExtensionsYml` throws `IOException`. Is there an intentional reason we are being overly broad in catching all exceptions here?

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 8]
```diff
@@ -67,84 +66,79 @@ public void testExtensionsDiscovery() throws Exception {
...
 
-        Set<DiscoveryExtension> expectedExtensionsSet = new HashSet<DiscoveryExtension>();
+        List<DiscoveryExtension> expectedExtensionsList = new ArrayList<DiscoveryExtension>();
 
-        expectedExtensionsSet.add(
+        expectedExtensionsList.add(
             new DiscoveryExtension(
                 "firstExtension",
-                "id",
+                "uniqueid1",
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:99]: Actually a serious comment!  Are ID's envisioned to be text-based like this, or in UUID format, or what?  The example here should probably mimic our vision.

**Review by dbwiddis — APPROVED** [review]
Approved with comments, some of them serious.

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java (hunk 7 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:158]: I thought I had changed that in my last PR, per your suggestion.  Looks like it got lost when I was rebasing and merging and I didn't catch it

**Comment by peterzhuamazon** [comment]
@ryanbogan I have merged the gradle check into this feature/extensions branchs.
Please rebase your PR with upstream and you would get gradle check.

Thanks.

**Comment by ryanbogan** [comment]
start gradle check

**Comment by owaiskazi19** [comment]
> start gradle check

This won't work anymore. As we have moved our gradle check to GHA :)

**Comment by ryanbogan** [comment]
> > start gradle check
> 
> This won't work anymore. As we have moved our gradle check to GHA :)


How can I run it again?

**Comment by peterzhuamazon** [comment]
```
Hi Ryan, in order to trigger again just push another commit or rerun the workflow in the check list.
The latter probably requires repo admin access, but rebase your branch and push again can be done by everyone.
```

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 9]
```diff
@@ -150,56 +150,56 @@ private void extensionsDiscovery() throws IOException {
...
+                                Boolean.parseBoolean(extension.hasNativeController())
+                            )
+                        )
+                    );
+                    logger.info("Loaded extension: " + extension);
+                } catch (IllegalArgumentException e) {
+                    logger.error(e.toString());
                 }
             }
+            if (!extensionsList.isEmpty()) {
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:191]: Can we have a test to test extensions present in the list?

**Pull request merged**